pudb/remote.py

```diff
@@ -105,27 +112,63 @@ def __init__(self, host=PUDB_RDB_HOST, port=PUDB_RDB_PORT,
                 codecs.getencoder("utf-8"),
                 codecs.getdecoder("utf-8"),
                 codecs.getreader("utf-8"),
-                codecs.getwriter("utf-8"))
+                codecs.getwriter("utf-8"),
+            )
         else:
             sock_file = codecs.StreamReaderWriter(
-                raw_sock_file,
-                codecs.getreader("utf-8"),
-                codecs.getwriter("utf-8"))
+                raw_sock_file, codecs.getreader("utf-8"), codecs.getwriter("utf-8")
+            )
 
         self._handle = sys.stdin = sys.stdout = sock_file
 
-        import telnetlib as tn
+        # nc negotiation doesn't support telnet options
+        if not reverse:
+            import telnetlib as tn
+
+            raw_sock_file.write(tn.IAC + tn.WILL + tn.SGA)
+            resp = raw_sock_file.read(3)
+            assert resp == tn.IAC + tn.DO + tn.SGA
 
-        raw_sock_file.write(tn.IAC + tn.WILL + tn.SGA)
-        resp = raw_sock_file.read(3)
-        assert resp == tn.IAC + tn.DO + tn.SGA
+            raw_sock_file.write(tn.IAC + tn.WILL + tn.ECHO)
+            resp = raw_sock_file.read(3)
+            assert resp == tn.IAC + tn.DO + tn.ECHO
 
-        raw_sock_file.write(tn.IAC + tn.WILL + tn.ECHO)
-        resp = raw_sock_file.read(3)
-        assert resp == tn.IAC + tn.DO + tn.ECHO
+        Debugger.__init__(
+            self, stdin=self._handle, stdout=self._handle, term_size=term_size
+        )
 
-        Debugger.__init__(self, stdin=self._handle, stdout=self._handle,
-                term_size=term_size)
+    def get_client(self, host, port, search_limit=100, reverse=False):
+        if reverse:
+            client, address = self.get_reverse_socket_client(host, port)
+            self.host, self.port = address
+            self.ident = "{0}:{1}".format(self.me, self.port)
+        else:
+            self._sock, conn_info = self.get_socket_client(
+                host, port, search_limit=search_limit,
+            )
+            self.host, self.port = conn_info
+            self.ident = "{0}:{1}".format(self.me, self.port)
+            self.say(BANNER.format(self=self))
+            client, address = self._sock.accept()
+        client.setblocking(1)
+        return client, (address, self.port)
+
+    def get_reverse_socket_client(self, host, port):
+        _sock = socket.socket()
+        try:
+            _sock.connect((host, port))
+            _sock.setblocking(1)
+        except socket.error as exc:
+            if exc.errno == errno.ECONNREFUSED:
+                raise Exception(CONN_REFUSED.format(self=self))
```

Comment:
When trying this (with just `set_trace(reverse=True)`, I got:
```
Traceback (most recent call last):
  File "debug_me.py", line 10, in <module>
    from pudb.remote import set_trace; set_trace(reverse=True)
  File "/home/andreas/work/pudb/pudb/remote.py", line 243, in set_trace
    return debugger(
  File "/home/andreas/work/pudb/pudb/remote.py", line 223, in debugger
    rdb = _current[0] = RemoteDebugger(
  File "/home/andreas/work/pudb/pudb/remote.py", line 98, in __init__
    self._client, (address, port) = self.get_client(
  File "/home/andreas/work/pudb/pudb/remote.py", line 142, in get_client
    client, address = self.get_reverse_socket_client(host, port)
  File "/home/andreas/work/pudb/pudb/remote.py", line 163, in get_reverse_socket_client
    raise Exception(CONN_REFUSED.format(self=self))
AttributeError: 'RemoteDebugger' object has no attribute 'remote_addr'
```

Reply:
You have to specify the host and port because it works as a client.
It'll better to raise ValueError for this situation.

Reply:
> You have to specify the host and port because it works as a client.

I suspect that, nine times out of 10, it'll be connecting to `localhost` and 9999.

> It'll better to raise ValueError for this situation.

Agreed. Could you do that?

Reply:
Sure. fixed in  https://github.com/inducer/pudb/pull/409/commits/88c3dd3aa1bed8a8a2191635e907eafa1bdb9784